{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "import numpy as np\n",
    "import collections\n",
    "import argparse\n",
    "from Bio import SeqIO\n",
    "from itertools import product"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def headerNAC(k, foutput):\n",
    "    global listTotal, caracteres\n",
    "    listTotal = []\n",
    "    file = open(foutput, 'a')\n",
    "    file.write('%s,' % ('nameseq'))\n",
    "    permsList = [''.join(str(i) for i in x) for x in product(caracteres, repeat=k)]\n",
    "    # print (permsList)\n",
    "    for perm in permsList:\n",
    "        # print (perm)\n",
    "        listTotal.append(perm)\n",
    "        file.write('%s,' % (str(perm)))\n",
    "    file.write('label')\n",
    "    file.write('\\n')\n",
    "    return\n",
    "\n",
    "\n",
    "def chunks(seq, win, step):\n",
    "    seqlen = len(seq)\n",
    "    for i in range(0, seqlen, step):\n",
    "        j = seqlen if i+win>seqlen else i+win\n",
    "        yield seq[i:j]\n",
    "        if j==seqlen: break\n",
    "    return        \n",
    "    \n",
    "\n",
    "def chunks_two(seq, win):\n",
    "    seqlen = len(seq)\n",
    "    for i in range(seqlen):\n",
    "        j = seqlen if i+win>seqlen else i+win\n",
    "        yield seq[i:j]\n",
    "        if j==seqlen: break\n",
    "    return\n",
    "\n",
    "            \n",
    "def file_record(name_seq, foutput, labelDataset):\n",
    "    file = open(foutput, 'a')\n",
    "    file.write('%s,' % (name_seq))\n",
    "    for x in probabilities:\n",
    "        # print (x)\n",
    "        file.write('%s,' % (str(x[1])))\n",
    "    file.write(labelDataset)\n",
    "    file.write('\\n')\n",
    "    print ('Recorded Sequence: %s' % (name_seq))\n",
    "    return\n",
    "    \n",
    "\n",
    "# Nucleic acid composition (NAC), Di-nucleotide composition (DNC), Tri-nucleotide composition (TNC)\n",
    "def nacSeq(k, foutput, finput, labelDataset):\n",
    "    global probabilities\n",
    "    headerNAC(k, foutput)\n",
    "    for seq_record in SeqIO.parse(finput, 'fasta'):\n",
    "        seq = seq_record.seq\n",
    "        seq = seq.upper()\t\n",
    "        name_seq = seq_record.name\n",
    "        probabilities = []\n",
    "        kmer = {}\n",
    "        totalWindows = (len(seq) - k) + 1 # (L - k + 1)\n",
    "        permsList = [''.join(str(i) for i in x) for x in product(caracteres, repeat=k)]\n",
    "        for key in permsList:\n",
    "            kmer[key] = 0\n",
    "        kmer = collections.OrderedDict(sorted(kmer.items()))\n",
    "        for subseq in chunks_two(seq, k):\n",
    "            if subseq in kmer:\n",
    "                # print(subseq)\n",
    "                kmer[subseq] = kmer[subseq] + 1\n",
    "            else:\n",
    "                kmer[subseq] = 1\n",
    "        for key, value in kmer.items():\n",
    "            # print (key)\n",
    "            # print (value)\n",
    "            probabilities.append([str(key), value/totalWindows])\n",
    "        file_record(name_seq, foutput, labelDataset)\n",
    "    return\n",
    "\n",
    "\n",
    "def ExtractionTechniques(finput, foutput, tech):\n",
    "    global listTotal, caracteres, complement\n",
    "    labelDataset = 'DNA'\n",
    "    # tech = str(args.type)\n",
    "    seq = 1\n",
    "    stepw = 1\n",
    "    if seq == 1:\n",
    "        caracteres = ['A', 'C', 'G', 'T']\n",
    "        complement = {'A': 'T','C': 'G','G': 'C','T': 'A'}\n",
    "    else:\n",
    "        caracteres = ['A', 'C', 'G', 'U']\n",
    "        complement = {'A': 'U','C': 'G','G': 'C','U': 'A'}\n",
    "    if tech == 'NAC' or tech == 'nac':\n",
    "        nacSeq(1, foutput, finput, labelDataset)\n",
    "    elif tech == 'DNC' or tech == 'dnc':\n",
    "        nacSeq(2, foutput, finput, labelDataset)\n",
    "    elif tech == 'TNC' or tech == 'tnc':\n",
    "        nacSeq(3, foutput, finput, labelDataset)\n",
    "#############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'caracteres' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mExtractionTechniques\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mres.fasta\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mres_1.csv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mdnc\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[4], line 90\u001b[0m, in \u001b[0;36mExtractionTechniques\u001b[0;34m(finput, foutput, tech)\u001b[0m\n\u001b[1;32m     88\u001b[0m     nacSeq(\u001b[38;5;241m1\u001b[39m, foutput, finput, labelDataset)\n\u001b[1;32m     89\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m tech \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDNC\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m tech \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdnc\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[0;32m---> 90\u001b[0m     \u001b[43mnacSeq\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfoutput\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfinput\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlabelDataset\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     91\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m tech \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTNC\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m tech \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtnc\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m     92\u001b[0m     nacSeq(\u001b[38;5;241m3\u001b[39m, foutput, finput, labelDataset)\n",
      "Cell \u001b[0;32mIn[4], line 50\u001b[0m, in \u001b[0;36mnacSeq\u001b[0;34m(k, foutput, finput, labelDataset)\u001b[0m\n\u001b[1;32m     48\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mnacSeq\u001b[39m(k, foutput, finput, labelDataset):\n\u001b[1;32m     49\u001b[0m     \u001b[38;5;28;01mglobal\u001b[39;00m probabilities\n\u001b[0;32m---> 50\u001b[0m     \u001b[43mheaderNAC\u001b[49m\u001b[43m(\u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfoutput\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     51\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m seq_record \u001b[38;5;129;01min\u001b[39;00m SeqIO\u001b[38;5;241m.\u001b[39mparse(finput, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfasta\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m     52\u001b[0m         seq \u001b[38;5;241m=\u001b[39m seq_record\u001b[38;5;241m.\u001b[39mseq\n",
      "Cell \u001b[0;32mIn[4], line 6\u001b[0m, in \u001b[0;36mheaderNAC\u001b[0;34m(k, foutput)\u001b[0m\n\u001b[1;32m      4\u001b[0m file \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(foutput, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124ma\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      5\u001b[0m file\u001b[38;5;241m.\u001b[39mwrite(\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m%\u001b[39m (\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnameseq\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[0;32m----> 6\u001b[0m permsList \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(\u001b[38;5;28mstr\u001b[39m(i) \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m x) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m product(\u001b[43mcaracteres\u001b[49m, repeat\u001b[38;5;241m=\u001b[39mk)]\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m# print (permsList)\u001b[39;00m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m perm \u001b[38;5;129;01min\u001b[39;00m permsList:\n\u001b[1;32m      9\u001b[0m     \u001b[38;5;66;03m# print (perm)\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'caracteres' is not defined"
     ]
    }
   ],
   "source": [
    "ExtractionTechniques(\"res.fasta\", 'res_1.csv', 'dnc')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PGAMenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
